src/azul/service/avro_pfb.py

```diff
@@ -0,0 +1,466 @@
+    pass_thru_uuid4,
+    value_and_unit,
+)
+from azul.types import (
+    JSON,
+    MutableJSON,
+)
+
+log = logging.getLogger(__name__)
+
+
+@attr.s(auto_attribs=True, frozen=True, kw_only=True)
+class PFBRelation:
+    dst_id: str
+    # A more appropriate attribute name would be dst_type, but we stick with
+    # 'dst_name' to conform to PFB spec
+    dst_name: str
+
+    @classmethod
+    def to_entity(cls, entity: 'PFBEntity'):
+        return cls(dst_id=entity.id, dst_name=entity.name)
+
+
+@attr.s(auto_attribs=True, frozen=True, kw_only=True)
+class PFBEntity:
+    """
+    Python representation of the PFB data object. Attribute names conform to the
+    PFB spec (which simplifies serialization).
+    """
+    namespace: ClassVar[uuid.UUID] = uuid.UUID(
+        '3372fc62-76d2-4ada-a68a-f08d4f558da5')
+    id: str
+    name: str
+    object: MutableJSON = attr.ib(eq=False)
+
+    @classmethod
+    def from_json(cls,
+                  name: str,
+                  object_: MutableJSON,
+                  schema: JSON) -> 'PFBEntity':
+        """
+        Derive ID from object in a reproducible way so that we can distinguish
+        entities by comparing their IDs.
+        """
+        # Entities need to be quickly distinguishable, but unfortunately
+        # document_id wasn't unique (or present) for all entities. Thus, we use
+        # the frozen state of the JSON to generate our hash and check equality.
+        entity_id = str(uuid.uuid5(cls.namespace, repr(sort_frozen(freeze(object_)))))
+        cls.add_missing_fields(name, object_, schema)
+        cls.replace_null_with_empty_string(object_)
+        return cls(id=entity_id, name=name, object=object_)
+
+    @classmethod
+    def add_missing_fields(cls, name: str, object_: MutableJSON, schema):
+        """
+        Compare entities against the schema and add any fields that are missing.
+
+        None is the default value, but because of https://github.com/DataBiosphere/azul/issues/2370
+        this isn't currently reflected in the schema.
+        """
+        object_schema = one(f for f in schema['fields'] if f['name'] == 'object')
+        entity_schema = one(e for e in object_schema['type'] if e['name'] == name)
+        for field in entity_schema['fields']:
+            if field['name'] not in object_:
+                if name == 'files':
+                    default_value = None
+                else:
+                    assert field['type']['type'] == 'array'
+                    # Default for a list of records is an empty list
+                    if isinstance(field['type']['items'], dict):
+                        assert field['type']['items']['type'] == 'record'
+                        default_value = []
+                    # All other types are lists of primitives where None is an accepted value
+                    else:
+                        # Change 'string' to 'null' in https://github.com/DataBiosphere/azul/issues/2462
```

Comment:
Use FIXME

And please review all ticket references. I don't want to continually stumble over these things in your contributions.